presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -46,6 +46,7 @@
     private final Optional<Compressor> compressor;
     private final Optional<Decompressor> decompressor;
     private final Optional<SpillCipher> spillCipher;
+    private byte[] compressionBuffer;
```

Comment:
I'm not sure this is a good idea. Since `compressionBuffer` grows as needed but never shrinks, serializing a single large Page or Slice can create a large heap allocation and keep it for the remaining lifespan of the `PagesSerde` instance.

Additionally, using `Slices.wrappedBuffer(compressionBuffer, 0, compressedSize)` without a subsequent call to `Slices.copyOf` means that the returned `Slice` can have its content overwritten by the next call to `serialize`.

If avoiding the slice copy and sharing the scratch buffer an important enough performance gain, I'd recommend having the caller manage the shared buffer and pass whatever buffer reuse abstraction is necessary as an argument. Only the caller can know when it's safe to reuse the buffer or at what point it's no longer needed.

Reply:
@pettyjamesm Petty, thanks for your comment. For the compressed case we should copy the slice when the compressed data is serialized. Thanks for pointing out and I added this logic. The copy happens only when the compression ratio is good.

However for the case where the original slice is used, we don't need to copy it because this function is only used by OptimizedPartitionedOutputOperator in which the slice is precisely sized when the SliceOutput was created. We used DynamicSliceOutput just to make the code look nicer, not that the memory need to be grown. We have already copied the slice for the compressed and spillCiphered case, and for the plain no-compression, no ciphiered case there is no need to copy the slice. Copying it incurs over 10% regression for the operator.